File: backend_nodejs/src/presentation/handler/markAsCompletedTodoHandler.ts
import type { NextFunction, Request, Response } from "express";
import z from "zod";
import type { MarkAsCompletedTodoUsecase } from "@/application_service/usecase/markAsCompletedTodoUsecase";
import { DomainConflictError } from "@/domain/model/errors";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import {
  BadRequestError,
  ConflictError,
  NotFoundError,
} from "@/presentation/middleware/errorHandler";

const markAsCompletedTodoRequestSchema = z.object({
  params: z.object({
    id: z.uuid({ version: "v7" }),
  }),
});

type MarkAsCompletedTodoResponse = {
  id: string;
  title: string;
  description: string | null;
  completed: boolean;
};

export class MarkAsCompletedTodoHandler {
  constructor(private markAsCompletedTodoUsecase: MarkAsCompletedTodoUsecase) {}

  handle = async (
    req: Request,
    res: Response<MarkAsCompletedTodoResponse>,
    next: NextFunction,
  ) => {
    try {
      const parsedReq = markAsCompletedTodoRequestSchema.parse(req);
      const result = await this.markAsCompletedTodoUsecase.execute({
        id: parsedReq.params.id,
      });
      res.json({
        id: result.todo.id,
        title: result.todo.title,
        description: result.todo.description,
        completed: result.todo.isCompleted,
      });
    } catch (error) {
      switch (true) {
        case error instanceof z.ZodError: {
          const message = error.issues
            .map((issue) => `${issue.path}: ${issue.message}`)
            .join(", ");
          return next(
            new BadRequestError(`Invalid request parameters: [${message}]`),
          );
        }
        case error instanceof RepositoryNotFoundError:
          return next(new NotFoundError("Todo not found"));
        case error instanceof DomainConflictError:
          return next(new ConflictError(error.message));
      }
      return next(error);
    }
  };
}

File: backend_python/src/todo_api/presentation/handler/mark_as_completed_todo_handler.py
from pydantic import BaseModel

from todo_api.application_service.usecase.mark_as_completed_todo_usecase import (
    MarkAsCompletedTodoUsecase,
    MarkAsCompletedTodoUsecaseInput,
)
from todo_api.domain.model.errors import TodoAlreadyCompletedError
from todo_api.domain.repository.errors import RepositoryNotFoundError
from todo_api.presentation.middleware.error_handler import ConflictError, NotFoundError
from todo_api.utils.uuid import UUID7


class MarkAsCompletedTodoResponse(BaseModel):
    id: str
    title: str
    description: str | None = None
    completed: bool


class MarkAsCompletedTodoHandler:
    def __init__(self, mark_as_completed_usecase: MarkAsCompletedTodoUsecase) -> None:
        self.mark_as_completed_usecase = mark_as_completed_usecase

    def handle(self, id: UUID7) -> MarkAsCompletedTodoResponse:
        try:
            result = self.mark_as_completed_usecase.execute(MarkAsCompletedTodoUsecaseInput(id=id))
        except RepositoryNotFoundError:
            raise NotFoundError(message="Todo not found")
        except TodoAlreadyCompletedError:
            raise ConflictError(message="Todo is already completed")
        return MarkAsCompletedTodoResponse(
            id=str(result.todo.id),
            title=result.todo.title,
            description=result.todo.description,
            completed=result.todo.completed,
        )
